### PR TITLE
Hw11

### PR DESCRIPTION
**Pull request by breengles**

### CODE REVIEW

File: utils/hw11_model.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+
+class BYOL(nn.Module):
+    def __init__(self, latent_dim=128) -> None:
+        super().__init__()
+        self.student = Net(1, latent_dim)
+        self.teacher = deepcopy(self.student)
+        self.teacher.requires_grad_(False)
+
+        self.transforms = transforms.Compose(
```
Comment (kraalfar) [inline utils/hw11_model.py:60]: добавление colorjitter должно улучшить accuracy скорее всего

File: utils/hw11_model.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    @property
+    def device(self):
+        return next(self.parameters()).device
+
+    @staticmethod
+    def C(z1, z2):
+        numerator = torch.einsum("bi,bj->ij", z1, z2)
+        denom = torch.sqrt((z1 ** 2).sum(0)) * torch.sqrt((z2 ** 2).sum(0)).reshape(-1, 1)  # reshape to get ij matrix
+        return numerator / denom / z1.shape[0]
```
Comment (kraalfar) [inline utils/hw11_model.py:160]: тут не надо было делить на размер батча или на denum, потому что это одно и тоже после нормализации

**Comment by kraalfar** [comment]
по первой имплементация ок, поэтому доделать не надо
по второй пока так. и код и ошибка очень на работу Семина похожи. вы одинаково отошли от псевдокода из статьи/лекции и одинаково ошиблись. поэтому пока считается списыванием

File: utils/hw11_model.py (hunk 2 above)
Comment (breengles) [reply utils/hw11_model.py:160]: хм, это было бы так, если бы мы нормировали на 1 по батчу, а мы же нормируем к 0 среднему и 1 стд, что, вообще говоря, не то же самое, разве нет? Да и в статье они делят...

**Comment by breengles** [comment]
Не понял, в каком месте отошел от псевдокода из статьи... буквально по нему делал + по формуле (2) для матрицы кросс-корреляции, в алгоритме 1 авторы статьи так же нормируют к 0 среднему и к 1 стд по батчу. Единственное, что отличается (но это не должно влиять) это то, что вместо вычитания каких-то единичных матриц я вычитаю из 1 диагональ, ну это буквально формула (1) для лосса из статьи...

**Comment by kraalfar** [comment]
вместо torch.mm einsum, я про это

**Comment by breengles** [comment]
ну так это же буквально числитель в формуле (2)
про torch.mm не знал, спасибо...
с другой стороны у торча и off_diagonal как-то нет :/

**Comment by breengles** [comment]
хм, кажется теперь понимаю, почему авторы делят только на N, я-то думал, этот mm это буквально формула (2), окей, но надо подумать, почему тогда нормировка так совпадет...

**Comment by kraalfar** [comment]
деление на стд это деление на корень из (суммы квадратов к размер батча) просто по определению стд для случайной величины с нулевым мо, поэтому это та же нормировка но в размер батча отличающаяся

**Comment by breengles** [comment]
дэ, спасибо, согласился

**Comment by breengles** [comment]
подправил, стало действительно лучше, спасибо. Лосс все равно колбасит, конечно, но теперь хотя бы он выглядит адекватнее + аккураси побольше стало.

по поводу списывания: не соглашусь, т.к., видимо, это неудачное стечение обстоятельств или банальное совпадение. Не знаю, что имено делал Кирилл, но статьи мы частично вместе разбирали. Предполагаю, что из-за этого и возникла такая ошибка у нас :\

File: utils/hw11_model.py (hunk 1 above)
Comment (kraalfar) [reply utils/hw11_model.py:60]: а он их же все равно меняет, потому что там преобразование не в rgb координатах
